File: 2024/Day05.cpp
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

#include "Days.hpp"
#include "utils.hpp"

static map<long, vector<long>> grules;

static bool comp(long a, long b) {
  if (grules.contains(a)) {
    if (find(grules[a].begin(), grules[a].end(), b) != grules[a].end()) {
      return true;
    }
  }

  if (grules.contains(b)) {
    if (find(grules[b].begin(), grules[b].end(), a) != grules[b].end()) {
      return false;
    }
  }

  return a < b;
}

Day05::Day05(string path) : AdventDay(path) {
  day = 5;
  title = "Print Queue";
}

bool Day05::isValid(const vector<long>& nums) {
  set<long> seen;
  for (auto n : nums) {
    // does this update number has grules?
    if (grules.contains(n)) {
      // check all the grules numbers
      for (auto r : grules[n]) {
        if (seen.contains(r)) {
          return false;
        }
      }
    }
    seen.insert(n);
  }
  return true;
}

long Day05::updates(const vector<string>& lines, bool workOnOutOfOrder) {
  long res = 0;

  bool isrule = true;
  grules.clear();

  for (const string& line : lines) {
    if (line.length() == 0) {
      isrule = false;
      continue;
    }

    // Parse input line into longs
    vector<long> nums = extractLongs(line);

    if (isrule) {
      grules[nums[0]].push_back(nums[1]);
    } else {
      if (workOnOutOfOrder) {
        if (isValid(nums) == false) {
          sort(nums.begin(), nums.end(), comp);
          res += nums[nums.size() >> 1];
        }
      } else {
        if (isValid(nums)) {
          res += nums[nums.size() >> 1];
        }
      }
    }
  }

  return res;
}

void Day05::part1(vector<string> lines) {
  printLongResult(updates(lines, false));
}

void Day05::part2(vector<string> lines) {
  printLongResult(updates(lines, true));
}

File: 2024/python/day05.py
from aoc import *
import functools

rules = {}

def isvalid(nums):
    seen = set()
    for n in nums:
        if n in rules:
            for r in rules[n]:
                if r in seen:
                    return False
        seen.add(n)
    return True

def compare(a, b):
    if a in rules:
        if b in rules[a]:
            return -1
    elif b in rules:
        if a in rules[b]:
            return 1
    return 0

def updates(lines, reorder):
    global rules

    res = 0
    rules = {}  # clean the global rules
    updates = []

    inrules = True
    for aline in lines:
        aline = aline.strip()
        if aline == '':
            inrules = False
            continue

        if inrules:
            [a,b] = [int(x) for x in aline.split('|')]
            if a in rules:
                rules[a].append(b)
            else:
                rules[a] = [b]
        else:
            updates.append([int(x) for x in aline.split(',')])

    for u in updates:
        if reorder:
            # Part 2
            if isvalid(u) == False:
                sortedu = sorted(u, key=functools.cmp_to_key(compare))
                res += sortedu[(int(len(sortedu)/2))]
        else:
            # Part 1
            if isvalid(u):
                res += u[(int(len(u)/2))]
    return res

def part1(lines):
    return updates(lines, False)

def part2(lines):
    return updates(lines, True)

AoCRunnerAll(5, 'Print Queue', part1, part2)